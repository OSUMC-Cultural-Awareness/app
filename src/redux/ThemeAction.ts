import { Store } from ".";

export const updateTheme = (
  theme: Store["theme"]
): { type: string; payload: Store["theme"] } => ({
  type: "UPDATE_THEME",
  payload: theme,
});
