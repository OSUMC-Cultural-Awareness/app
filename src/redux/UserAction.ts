import { Store } from ".";

export const updateUser = (
  user: Store["user"]
): { type: string; payload: Store["user"] } => ({
  type: "UPDATE_USER",
  payload: user,
});

export const resetUser = (): {
  type: string;
  payload: Record<string, unknown>;
} => ({
  type: "RESET_USER",
  payload: {},
});
