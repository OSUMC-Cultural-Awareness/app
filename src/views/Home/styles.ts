import { StyleSheet, Platform } from "react-native";

import { darkTheme, lightTheme } from "../../theme";

export default StyleSheet.create({
  spinner: { top: "50%", position: "relative" },

  view: {
    flex: 1,
  },

  modalLight: {
    margin: 40,
    padding: 40,
    backgroundColor: lightTheme.colors.background,
  },

  modalDark: {
    margin: 40,
    padding: 40,
    backgroundColor: darkTheme.colors.background,
  },

  list: {
    flex: 1,
  },

  div: {
    margin: "5px",
  },

  deleteButton: {
    backgroundColor: "red",
    margin: 5,
  },

  fab: {
    // HACK: Any here so that FABs are Fixed on web, but use the default styling on Mobile.
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    position: (Platform.OS === "web" ? "fixed" : "absolute") as any,
    margin: 16,
    right: 0,
    bottom: 0,
  },

  ListActions: {
    flexDirection: "row",
  },

  editModalInput: {
    paddingBottom: 10,
  },

  cultureListActions: {
    flex: 1,
    flexDirection: "row",
    justifyContent: "flex-end",
  },

  dialogButton: {
    margin: 5,
  },
});
