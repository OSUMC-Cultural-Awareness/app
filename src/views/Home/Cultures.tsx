import React, { useState } from "react";
import { View, FlatList } from "react-native";
import { StackNavigationProp } from "@react-navigation/stack";
import { ActivityIndicator, List, IconButton } from "react-native-paper";

import { Routes } from "../../routes";

import { Culture, Ledger } from "../../lib";

import styles from "./styles";

/**
 * Properties for {@link Cultures}
 */
type CultureProps = {
  navigation: StackNavigationProp<Routes, "Home">;
  token: string;
  cultures: IterableIterator<[string, number]>;
  onRefresh: () => void;
  searchQuery?: string;
  offline: boolean;
};

/**
 * Component that displays a list of components of either {@link Cultures}
 *
 * @param {CultureProps} props
 * @returns {React.ReactElement} React component
 */
export default function Cultures(props: CultureProps): React.ReactElement {
  const { cultures, onRefresh, token, searchQuery, offline } = props;
  const [refreshing, setRefreshing] = useState(false);

  if (!cultures) {
    return (
      <ActivityIndicator animating={true} size="large" style={styles.spinner} />
    );
  }

  const deleteCulture = async (name: string) => {
    try {
      await Culture.delete(name, token);
    } catch (err) {
      console.error("Failed to delete culture", err);
    }
    onRefresh();
  };

  const searchResults = (): [string, number][] => {
    return Array.from(cultures).filter((culture) => {
      const [name] = culture;
      if (!searchQuery) {
        return true;
      }

      const nameLower = name.toLowerCase();
      const query = searchQuery.toLowerCase();

      return nameLower.includes(query);
    });
  };

  return (
    <View>
      <FlatList
        data={searchResults()}
        keyExtractor={(_, index) => index.toString()}
        onRefresh={() => {
          setRefreshing(true);
          onRefresh();
          setRefreshing(false);
        }}
        refreshing={refreshing}
        renderItem={({ item }) => {
          const [name] = item;
          return (
            <List.Item
              title={name}
              onPress={() =>
                props.navigation.navigate("Culture", { cultureName: name })
              }
              right={() => (
                <View
                  style={{
                    flex: 1,
                    flexDirection: "row",
                    justifyContent: "flex-end",
                  }}
                >
                  {!offline && (
                    <IconButton
                      icon="download"
                      onPress={() => Ledger.add(name)}
                    />
                  )}
                  {token !== "" && (
                    <IconButton
                      icon="delete"
                      onPress={() => deleteCulture(name)}
                    />
                  )}
                </View>
              )}
            />
          );
        }}
      />
    </View>
  );
}
