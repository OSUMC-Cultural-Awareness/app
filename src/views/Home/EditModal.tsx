import React, { useRef, useEffect } from "react";
import { useFormik } from "formik";
import { Dialog, TextInput, Button, useTheme } from "react-native-paper";

import { Admin } from "../../lib";

import { EmailNameValidation } from "./validation";
import styles from "./styles";

/**
 * Invite Email screen fields for Formik.
 */
type EditFields = {
  email: string;
  name: string;
};

/**
 * Initial values for email field for Formik.
 */
const initialValues: EditFields = {
  // This field could be updated with useEffect to enter the user's saved email address.
  email: "",
  name: "",
};

type Props = {
  token: string;
  show: boolean;
  admin: Admin;
  onDismiss: () => void;
  onErr: (err: string) => void;
  onRefresh: () => void;
};

export default function EditModal(props: Props): React.ReactElement {
  const { admin, token, show, onDismiss, onErr, onRefresh } = props;
  const name = useRef();
  const theme = useTheme();

  const {
    values,
    handleChange,
    handleBlur,
    errors,
    touched,
    handleSubmit,
    validateForm,
    setFieldValue,
  } = useFormik({
    validationSchema: EmailNameValidation,
    initialValues: initialValues,
    onSubmit: (values) => update(values),
  });

  useEffect(() => {
    setFieldValue("email", admin?.email);
    setFieldValue("name", admin?.name);
  }, [admin]);

  const update = async (fields: EditFields) => {
    const { name, email } = fields;

    await validateForm();

    try {
      await Admin.update(email, name, token);
      onDismiss();
      onRefresh();
    } catch (err) {
      onErr(err.toString());
    }
  };

  return (
    <Dialog
      visible={show}
      onDismiss={onDismiss}
      style={{ backgroundColor: theme.colors.surface }}
    >
      <Dialog.Title>Edit {admin?.name}</Dialog.Title>
      <Dialog.Content>
        <TextInput
          mode="outlined"
          style={{ paddingBottom: 10 }}
          left={<TextInput.Icon name="email" />}
          label="email"
          value={values.email}
          disabled={true}
        />
        <TextInput
          autoFocus={true}
          textContentType="name"
          mode="outlined"
          left={<TextInput.Icon name="account" />}
          error={errors.name && touched.name}
          onSubmitEditing={handleSubmit}
          returnKeyType="done"
          label="name"
          value={values.name}
          ref={name}
          onBlur={handleBlur("name")}
          onChangeText={handleChange("name")}
        />
      </Dialog.Content>
      <Dialog.Actions>
        <Button style={styles.dialogButton} onPress={onDismiss}>
          Cancel
        </Button>
        <Button
          style={styles.dialogButton}
          mode="contained"
          onPress={handleSubmit}
        >
          Save
        </Button>
      </Dialog.Actions>
    </Dialog>
  );
}
