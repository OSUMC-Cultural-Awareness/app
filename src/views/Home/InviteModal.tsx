import React, { useRef } from "react";
import { useFormik } from "formik";
import { Dialog, TextInput, Button } from "react-native-paper";
import { useTheme } from "react-native-paper";

import { Admin } from "../../lib";

import { EmailValidation } from "./validation";
import styles from "./styles";

/**
 * Invite Email screen fields for Formik.
 */
type EmailField = {
  email: string;
};

/**
 * Initial values for email field for Formik.
 */
const initialValues: EmailField = {
  // This field could be updated with useEffect to enter the user's saved email address.
  email: "",
};

type Props = {
  token: string;
  show: boolean;
  onDismiss: () => void;
  onErr: (err: string) => void;
};

export default function InviteModal(props: Props): React.ReactElement {
  const { token, show, onDismiss, onErr } = props;
  const theme = useTheme();

  const email = useRef();
  const {
    values,
    handleChange,
    handleBlur,
    errors,
    touched,
    handleSubmit,
    validateField,
  } = useFormik({
    validationSchema: EmailValidation,
    initialValues: initialValues,
    onSubmit: (values) => invite(values),
  });

  const invite = async (field: EmailField) => {
    const { email } = field;
    await validateField("email");

    try {
      await Admin.invite(email, token);
      onDismiss();
    } catch (err) {
      onErr(err.toString());
    }
  };

  return (
    <Dialog
      visible={show}
      onDismiss={onDismiss}
      style={{ backgroundColor: theme.colors.surface }}
    >
      <Dialog.Title>Invite a new Admin</Dialog.Title>
      <Dialog.Content>
        <TextInput
          textContentType="emailAddress"
          autoFocus={true}
          mode="outlined"
          onSubmitEditing={handleSubmit}
          returnKeyType="send"
          left={<TextInput.Icon name="email" />}
          error={errors.email && touched.email}
          label="email"
          value={values.email}
          ref={email}
          onBlur={handleBlur("email")}
          onChangeText={handleChange("email")}
        />
      </Dialog.Content>
      <Dialog.Actions>
        <Button style={styles.dialogButton} onPress={onDismiss}>
          Cancel
        </Button>
        <Button
          mode="contained"
          style={styles.dialogButton}
          onPress={handleSubmit}
        >
          Invite
        </Button>
      </Dialog.Actions>
    </Dialog>
  );
}
