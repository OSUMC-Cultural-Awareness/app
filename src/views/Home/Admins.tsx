import React, { useState } from "react";
import { FlatList, View } from "react-native";
import { connect } from "react-redux";
import {
  List,
  IconButton,
  Dialog,
  Portal,
  Button,
  useTheme,
} from "react-native-paper";

import { Store } from "../../redux";
import { Admin } from "../../lib";

import EditModal from "./EditModal";
import styles from "./styles";

/**
 * Properties for {@link Admins}
 */
type Props = {
  token: string;
  admins: Admin[];
  theme: string;
  user: Admin;
  onRefresh: () => void;
  searchQuery?: string;
  onMsg: (msg: string) => void;
};

/**
 * Component that displays a list of components of {@link Admin}
 *
 * @param {Props} props
 * @returns {React.ReactElement} React component
 */
function Admins(props: Props): React.ReactElement {
  const { user, token, admins, onRefresh, searchQuery, onMsg } = props;
  const theme = useTheme();

  const [deleteModal, setDeleteModal] = useState(false);
  const [editModal, setEditModal] = useState(false);
  const [selectedAdmin, setSelectedAdmin] = useState(null);

  const onDelete = async () => {
    try {
      await Admin.delete(selectedAdmin.email, token);
      onRefresh();
    } catch (err) {
      onMsg(err.toString());
    }
  };

  const searchResults = (): Admin[] => {
    return admins.filter((admin) => {
      if (!searchQuery) {
        return true;
      }

      const name = admin.email.toLowerCase();
      const query = searchQuery.toLowerCase();

      return name.includes(query);
    });
  };

  return (
    <FlatList
      style={styles.list}
      data={searchResults()}
      keyExtractor={(_, index) => index.toString()}
      renderItem={({ item }) => {
        return (
          <View>
            <List.Item
              title={item.email}
              right={() =>
                props.token !== "" && (
                  <View style={styles.ListActions}>
                    <IconButton
                      icon="pencil"
                      onPress={() => {
                        setEditModal(true);
                        setSelectedAdmin(item);
                      }}
                    />
                    {!item.superUser && item.email != user.email && (
                      <IconButton
                        icon="delete"
                        onPress={() => {
                          setDeleteModal(true);
                          setSelectedAdmin(item);
                        }}
                      />
                    )}
                  </View>
                )
              }
            />
            <Portal>
              <Dialog
                visible={deleteModal}
                onDismiss={() => setDeleteModal(false)}
                style={{ backgroundColor: theme.colors.surface }}
              >
                <Dialog.Title>
                  Are you sure you want to delete {selectedAdmin?.email}?
                </Dialog.Title>
                <Dialog.Actions>
                  <Button
                    style={styles.dialogButton}
                    onPress={() => setDeleteModal(false)}
                  >
                    Cancel
                  </Button>
                  <Button
                    mode="contained"
                    onPress={() => {
                      onDelete();
                      setDeleteModal(false);
                    }}
                    style={styles.deleteButton}
                  >
                    Delete
                  </Button>
                </Dialog.Actions>
              </Dialog>
              <EditModal
                show={editModal}
                token={token}
                admin={selectedAdmin}
                onDismiss={() => setEditModal(false)}
                onErr={onMsg}
                onRefresh={onRefresh}
              />
            </Portal>
          </View>
        );
      }}
    />
  );
}

export default connect(
  (state: Store) => ({
    user: state.user.user,
    theme: state.theme,
  }),
  null
)(Admins);
