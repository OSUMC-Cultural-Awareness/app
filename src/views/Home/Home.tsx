import React, { useState, useEffect, useRef } from "react";
import { View, Platform } from "react-native";

import { connect } from "react-redux";
import { StackNavigationProp } from "@react-navigation/stack";
import { RouteProp } from "@react-navigation/native";
import { createMaterialTopTabNavigator } from "@react-navigation/material-top-tabs";
import { getFocusedRouteNameFromRoute } from "@react-navigation/native";
import {
  Portal,
  Modal,
  Text,
  ActivityIndicator,
  TextInput,
  Button,
  FAB,
  Snackbar,
} from "react-native-paper";
import { useFormik } from "formik";

import { Store } from "../../redux";
import { Admin, Culture, OfflineError, Ledger } from "../../lib";
import { Routes } from "../../routes";

import Cultures from "./Cultures";
import Admins from "./Admins";
import styles from "./styles";
import { EmailValidation } from "./validation";

type Props = {
  navigation: StackNavigationProp<Routes, "Home">;
  route: RouteProp<Routes, "Home">;
  user: Admin;
  token: string;
  theme: string;
};

type TabProps = {
  Cultures: { cultures: Culture[] };
  Admins: { admins: Admin[] };
};

/**
 * Invite Email screen fields for Formik.
 */
type EmailField = {
  email: string;
};

/**
 * Initial values for email field for Formik.
 */
const initialValues: EmailField = {
  // This field could be updated with useEffect to enter the user's saved email address.
  email: "",
};

const Tab = createMaterialTopTabNavigator<TabProps>();

function Home(props: Props): React.ReactElement {
  const { token, route, navigation, user, theme } = props;

  const [cultures, setCultures] = useState(null);
  const [admins, setAdmins] = useState(null);
  const [inviteModal, setInviteModal] = React.useState(false);
  const [msg, setMsg] = useState<string>("");
  const [offline, setOffline] = useState(false);

  const email = useRef();

  const {
    values,
    handleChange,
    handleBlur,
    errors,
    touched,
    handleSubmit,
    validateField,
  } = useFormik({
    validationSchema: EmailValidation,
    initialValues: initialValues,
    onSubmit: (values) => invite(values),
  });

  const fetchCultures = async () => {
    try {
      const cultures = await Culture.list();
      setCultures(cultures);
    } catch (err) {
      if (err instanceof OfflineError) {
        try {
          const ledger = await Ledger.list();
          let cultures = [];
          ledger.forEach((_, key) =>
            cultures.push({ name: key, modified: cultures[key] })
          );
          setCultures(cultures);
          setOffline(true);
        } catch (err) {
          setMsg(err.toString());
        }
      } else {
        setMsg(err.toString());
      }
    }
  };

  useEffect(() => {
    fetchCultures();
  }, []);

  const fetchAdmins = async () => {
    if (!props.token) {
      return;
    }

    const admins = user.superUser ? await Admin.list(token) : [user];
    setAdmins(admins);
  };

  useEffect(() => {
    fetchAdmins();
  }, []);

  if (!token) {
    return (
      <Cultures
        navigation={props.navigation}
        token={""}
        cultures={cultures}
        onRefresh={() => fetchCultures()}
        offline={offline}
      />
    );
  }

  const onAdd = () => {
    switch (getFocusedRouteNameFromRoute(route) ?? "Cultures") {
      case "Cultures":
        setCultures([
          ...cultures,
          { name: "New Culture", modified: Date.now() },
        ]);
        break;
      case "Admins":
        setInviteModal(true);
    }
  };

  const invite = async (field: EmailField) => {
    const { email } = field;
    await validateField("email");
    try {
      await Admin.invite(email, token);
      setInviteModal(false);
    } catch (err) {
      setMsg(err.toString());
    }
  };

  const hideSnackbar = () => setMsg("");

  const styleFAB = {
    position: Platform.OS === "web" ? "fixed" : "absolute",
    margin: 16,
    right: 0,
    bottom: 0,
  };

  if (!admins) {
    return (
      <ActivityIndicator animating={true} size="large" style={styles.spinner} />
    );
  }

  return (
    <View style={styles.view}>
      <Tab.Navigator initialRouteName="Cultures">
        <Tab.Screen name="Cultures">
          {() => (
            <Cultures
              navigation={navigation}
              token={token}
              cultures={cultures}
              onRefresh={() => fetchCultures()}
              offline={offline}
            />
          )}
        </Tab.Screen>
        <Tab.Screen name="Admins">
          {() => (
            <Admins
              token={token}
              admins={admins}
              onRefresh={() => fetchAdmins()}
            />
          )}
        </Tab.Screen>
      </Tab.Navigator>
      <FAB style={styleFAB as any} icon="plus" onPress={onAdd} />
      <Portal>
        <Snackbar
          visible={msg != ""}
          onDismiss={() => setMsg("")}
          action={{
            label: "Undo",
            onPress: () => setMsg(""),
          }}
        >
          {msg}
        </Snackbar>
        <Modal
          visible={inviteModal}
          contentContainerStyle={
            theme === "Dark" ? styles.modalDark : styles.modalLight
          }
          onDismiss={() => setInviteModal(false)}
        >
          <Text>Invite a new admin</Text>
          <TextInput
            autoFocus={true}
            textContentType="emailAddress"
            mode="outlined"
            left={<TextInput.Icon name="email" />}
            error={errors.email && touched.email}
            label="email"
            value={values.email}
            ref={email}
            onBlur={handleBlur("email")}
            onChangeText={handleChange("email")}
          />
          <View style={styles.div} />
          <Button mode="contained" onPress={handleSubmit}>
            Send Invite
          </Button>
        </Modal>
      </Portal>
      <Portal>
        <Snackbar
          visible={msg !== ""}
          onDismiss={hideSnackbar}
          action={{
            label: "Ok",
            onPress: hideSnackbar,
          }}
        >
          {msg}
        </Snackbar>
      </Portal>
    </View>
  );
}

export default connect(
  (
    state: Store,
    ownProps: {
      navigation: StackNavigationProp<Routes, "Home">;
      route: RouteProp<Routes, "Home">;
    }
  ) => ({
    navigation: ownProps.navigation,
    route: ownProps.route,
    user: state.user.user,
    token: state.user.token,
    theme: state.theme,
  }),
  null
)(Home);
