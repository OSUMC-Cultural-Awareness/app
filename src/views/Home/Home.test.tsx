/**
 * @jest-environment jsdom
 */
import React from "react";
import "react-native";
import { NavigationContainer } from "@react-navigation/native";
import { createStackNavigator } from "@react-navigation/stack";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { shallow } from "enzyme";

import Home from ".";
import { USER_INITIAL_STATE, THEME_INITIAL_STATE } from "../../redux";
import { Routes } from "../../routes";

jest.useFakeTimers();

const HomeTest = () => {
  const Stack = createStackNavigator<Routes>();

  const mockStore = configureStore([]);

  return (
    <Provider
      store={mockStore({
        user: USER_INITIAL_STATE,
        theme: THEME_INITIAL_STATE,
      })}
    >
      <NavigationContainer>
        <Stack.Navigator initialRouteName="Home">
          <Stack.Screen name="Home" component={Home} />
        </Stack.Navigator>
      </NavigationContainer>
    </Provider>
  );
};

describe("Home", () => {
  test("Renders", () => {
    expect(shallow(<HomeTest />)).toMatchSnapshot();
  });
});
