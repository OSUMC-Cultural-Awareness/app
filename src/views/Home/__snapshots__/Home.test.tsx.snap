// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Renders 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <ForwardRef(NavigationContainer)>
    <StackNavigator
      initialRouteName="Home"
    >
      <Screen
        component={
          Object {
            "$$typeof": Symbol(react.memo),
            "WrappedComponent": [Function],
            "compare": null,
            "displayName": "Connect(Home)",
            "type": [Function],
          }
        }
        name="Home"
      />
    </StackNavigator>
  </ForwardRef(NavigationContainer)>
</Provider>
`;
